CUDADataFormats/Track/interface/TrackSoAHeterogeneousT.h

```diff
@@ -100,7 +117,99 @@ namespace pixelTrack {
 
   using TrackSoA = TrackSoAHeterogeneousT<maxNumber()>;
   using TrajectoryState = TrajectoryStateSoAT<maxNumber()>;
-  using HitContainer = TrackSoA::HitContainer;
+  // using HitContainer = TrackSoA::HitContainer;
+
+  template <typename TrackerTraits>
+  using TrackSoAT = PixelTrackSoAT<TrackerTraits>;
+
+  template <typename TrackerTraits>
+  using HitContainerT = typename PixelTrackSoAT<TrackerTraits>::HitContainer;
+
+  using TrackSoAPhase1 = TrackSoAT<pixelTopology::Phase1>;
+  using TrackSoAPhase2 = TrackSoAT<pixelTopology::Phase2>;
+
+  template <typename TrackerTraits, typename Enable = void>
+  struct QualityCutsT {};
+
+  template <typename TrackerTraits>
+  struct QualityCutsT<TrackerTraits, isPhase1Topology<TrackerTraits>> {
+    // chi2 cut = chi2Scale * (chi2Coeff[0] + pT/GeV * (chi2Coeff[1] + pT/GeV * (chi2Coeff[2] + pT/GeV * chi2Coeff[3])))
+    float chi2Coeff[4];
+    float chi2MaxPt;  // GeV
+    float chi2Scale;
+
+    struct Region {
+      float maxTip;  // cm
+      float minPt;   // GeV
+      float maxZip;  // cm
+    };
+
+    Region triplet;
+    Region quadruplet;
+
+    __device__ __forceinline__ bool isHP(PixelTrackSoAT<TrackerTraits> const *__restrict__ tracks,
+                                         int nHits,
+                                         int it) const {
+      // impose "region cuts" based on the fit results (phi, Tip, pt, cotan(theta)), Zip)
+      // default cuts:
+      //   - for triplets:    |Tip| < 0.3 cm, pT > 0.5 GeV, |Zip| < 12.0 cm
+      //   - for quadruplets: |Tip| < 0.5 cm, pT > 0.3 GeV, |Zip| < 12.0 cm
+      // (see CAHitNtupletGeneratorGPU.cc)
+      auto const &region = (nHits > 3) ? quadruplet : triplet;
+      return (std::abs(tracks->tip(it)) < region.maxTip) and (tracks->pt(it) > region.minPt) and
+             (std::abs(tracks->zip(it)) < region.maxZip);
+    }
+
+    __device__ __forceinline__ bool strictCut(PixelTrackSoAT<TrackerTraits> const *__restrict__ tracks, int it) const {
+      auto roughLog = [](float x) {
+        // max diff [0.5,12] at 1.25 0.16143
+        // average diff  0.0662998
+        union IF {
+          uint32_t i;
+          float f;
+        };
+        IF z;
```

Comment:
is a union allowed in this context? @makortel

Reply:
It is undefined behavior (as elsewhere). Discussion in 
https://stackoverflow.com/questions/47037104/cuda-type-punning-memcpy-vs-ub-union
suggests that the current implementation of `edm::bit_cast` could add overhead in CUDA device code. One option would be to amend `edm::bit_cast` as suggested there (and double-check that the actual `memcpy()` ends up elided).

Alternatively this discussion
https://forums.developer.nvidia.com/t/packed-datatype-in-an-union-volatile-keyword-necessary/45054
points to [`__float_as_uint()`](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__CAST.html#group__CUDA__MATH__INTRINSIC__CAST_1gaf4c4e5365416bb96b5937df9ffd9497) CUDA intrinsic. So one could have `#ifdef` here between that (for CUDA) and `edm::bit_cast` (for host code). And hopefully we could soon (before we need to think any other platform) just move to `std::bit_cast`.

Reply:
This one doesn't seem to be resolved? @slava77 can you clarify?

Reply:
Given this is how it's implemented already now ([here](https://cmssdt.cern.ch/dxr/CMSSW/source/RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorKernelsImpl.h#452)) I thought it was more of a general comment rather than specific to this PR. I can take care in the followin-up PRs I want to open connected to this.